File: win-service-manager/agent/core/config_merger.py
import xml.etree.ElementTree as ET
from pathlib import Path

def merge_dotnet_config(existing_config: Path, release_config: Path, target_output: Path) -> None:
    """
    Executa o Smart Merge de arquivos .config do .NET.
    Garante que parâmetros locais da máquina (chaves Zabbix, paths locais e endpoints)
    existentes no .config antigo não sejam sobrescritos pelo .config novo da release.
    Novas seções ou chaves trazidas na release são preservadas intactas.
    """
    existing_config = Path(existing_config)
    release_config = Path(release_config)
    target_output = Path(target_output)

    old_settings = {}
    old_app_settings = {}

    # 1. Mapear configurações do arquivo existente
    if existing_config.exists():
        try:
            old_tree = ET.parse(existing_config)
            old_root = old_tree.getroot()

            # Mapear applicationSettings e userSettings existentes
            for s in old_root.findall(".//setting"):
                name = s.get("name")
                val = s.find("value")
                if name and val is not None:
                    old_settings[name] = val.text

            # Mapear appSettings existentes
            for a in old_root.findall(".//appSettings/add"):
                k = a.get("key")
                v = a.get("value")
                if k:
                    old_app_settings[k] = v
        except Exception as e:
            print(f"[config_merger] Aviso: Erro ao ler config existente {existing_config}: {e}")

    # 2. Carregar o novo arquivo de configuração da release
    if not release_config.exists():
        raise FileNotFoundError(f"Arquivo de release .config não encontrado em {release_config}")

    new_tree = ET.parse(release_config)
    new_root = new_tree.getroot()

    # 3. Injetar valores salvos de applicationSettings / userSettings no novo XML
    for s in new_root.findall(".//setting"):
        name = s.get("name")
        if name in old_settings:
            val = s.find("value")
            if val is not None:
                val.text = old_settings[name]

    # 4. Injetar valores salvos de appSettings no novo XML
    for a in new_root.findall(".//appSettings/add"):
        k = a.get("key")
        if k in old_app_settings:
            a.set("value", old_app_settings[k])

    # 5. Salvar arquivo mesclado garantindo declaração UTF-8
    target_output.parent.mkdir(parents=True, exist_ok=True)
    new_tree.write(target_output, encoding="utf-8", xml_declaration=True)
    print(f"[config_merger] Smart Merge concluído com sucesso: {target_output}")

File: win-service-manager-dotnet/WinServiceFleetAgent/Core/ConfigMerger.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Xml.Linq;

namespace WinServiceFleetAgent.Core
{
    public static class ConfigMerger
    {
        public static void MergeDotNetConfig(string existingConfigPath, string releaseConfigPath, string targetOutputPath)
        {
            var oldSettings = new Dictionary<string, string>();
            var oldAppSettings = new Dictionary<string, string>();

            // 1. Mapear configurações do arquivo existente
            if (File.Exists(existingConfigPath))
            {
                try
                {
                    var oldDoc = XDocument.Load(existingConfigPath);
                    foreach (var s in oldDoc.Descendants("setting"))
                    {
                        var name = s.Attribute("name")?.Value;
                        var val = s.Element("value")?.Value;
                        if (!string.IsNullOrEmpty(name) && val != null)
                        {
                            oldSettings[name] = val;
                        }
                    }

                    foreach (var appSetting in oldDoc.Descendants("appSettings").Descendants("add"))
                    {
                        var key = appSetting.Attribute("key")?.Value;
                        var val = appSetting.Attribute("value")?.Value;
                        if (!string.IsNullOrEmpty(key) && val != null)
                        {
                            oldAppSettings[key] = val;
                        }
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"[ConfigMerger] Erro ao ler config existente '{existingConfigPath}': {ex.Message}");
                }
            }

            // 2. Carregar o arquivo novo trazido pela release
            if (!File.Exists(releaseConfigPath))
            {
                throw new FileNotFoundException($"Arquivo de release config não encontrado em: {releaseConfigPath}");
            }

            var newDoc = XDocument.Load(releaseConfigPath);

            // 3. Injetar valores salvos de applicationSettings / userSettings
            foreach (var s in newDoc.Descendants("setting"))
            {
                var name = s.Attribute("name")?.Value;
                if (!string.IsNullOrEmpty(name) && oldSettings.TryGetValue(name, out var oldVal))
                {
                    var valElem = s.Element("value");
                    if (valElem != null)
                    {
                        valElem.Value = oldVal;
                    }
                }
            }

            // 4. Injetar valores salvos de appSettings
            foreach (var appSetting in newDoc.Descendants("appSettings").Descendants("add"))
            {
                var key = appSetting.Attribute("key")?.Value;
                if (!string.IsNullOrEmpty(key) && oldAppSettings.TryGetValue(key, out var oldVal))
                {
                    appSetting.SetAttributeValue("value", oldVal);
                }
            }

            // 5. Salvar XML mesclado
            var targetDir = Path.GetDirectoryName(targetOutputPath);
            if (!string.IsNullOrEmpty(targetDir) && !Directory.Exists(targetDir))
            {
                Directory.CreateDirectory(targetDir);
            }

            newDoc.Save(targetOutputPath);
            Console.WriteLine($"[ConfigMerger] Smart Merge concluído com sucesso: {targetOutputPath}");
        }
    }
}
